src/fontra/views/editor/editor.js

```diff
@@ -1780,6 +1847,174 @@ export class EditorController {
     });
   }
 
+  canAddAnchor() {
+    return this.sceneModel.getSelectedPositionedGlyph()?.glyph.canEdit;
+  }
+
+  async doAddAnchor() {
+    const point = this.sceneController.selectedGlyphPoint(this.contextMenuPosition);
+    const { anchor: tempAnchor } = await this.doAddEditAnchorDialog(undefined, point);
+    if (!tempAnchor) {
+      return;
+    }
+
+    const newAnchor = {
+      name: tempAnchor.name ? tempAnchor.name : "anchorName",
+      // x will be specified individually for each layer
+      y: tempAnchor.y ? tempAnchor.y : Math.round(point.y),
+    };
+    const relativeScaleX =
+      Math.round(point.x) /
+      this.sceneModel.getSelectedPositionedGlyph().glyph.instance.xAdvance;
+
+    await this.sceneController.editLayersAndRecordChanges((layerGlyphs) => {
+      for (const layerGlyph of Object.values(layerGlyphs)) {
+        const anchorX = tempAnchor.x
+          ? tempAnchor.x
+          : layerGlyph.xAdvance * relativeScaleX;
+        layerGlyph.anchors.push({ name: newAnchor.name, x: anchorX, y: newAnchor.y });
+      }
+      const instance = this.sceneModel.getSelectedPositionedGlyph().glyph.instance;
+      const newAnchorIndex = instance.anchors.length - 1;
+      this.sceneController.selection = new Set([`anchor/${newAnchorIndex}`]);
+      return "Add Anchor";
+    });
+  }
+
+  async doAddEditAnchorDialog(anchor = undefined, point = undefined) {
+    const titlePrefix = anchor ? "Edit" : "Add";
+    if (!anchor && !point) {
+      // Need at least one of the two
+      return {};
+    }
+
+    const instance = this.sceneModel.getSelectedPositionedGlyph().glyph.instance;
+
+    const validateInput = () => {
+      const warnings = [];
+      const editedAnchorName =
+        nameController.model.anchorName || nameController.model.suggestedAnchorName;
+      if (!editedAnchorName.length) {
+        warnings.push("⚠️ The name must not be empty");
+      }
+      if (
+        !(
+          nameController.model.anchorName ||
+          nameController.model.anchorX ||
+          nameController.model.anchorY
+        )
+      ) {
+        warnings.push("⚠️ At least one element need to be specified");
```

Comment:
I don't understand this warning. I think the "Edit" dialog should just be completely populated, there is no need for placeholder magic at all here.

Reply:
When I double-click an anchor, I get this dialog:
<img width="396" alt="image" src="https://github.com/googlefonts/fontra/assets/4246121/4f107c8c-aa31-49a8-a692-2f5858d9f044">
This warning is only confusing.

Reply:
Btw if I double-click an anchor, and fill in any value, there is an error in the console that's coming from the server:
<img width="1108" alt="image" src="https://github.com/googlefonts/fontra/assets/4246121/5d685b3c-e3b8-4589-ae9e-525d42fe0bcd">

The server error suggests you are sending a value that is not a float or an int:
```
2024-04-26 12:27:31 fontra.core.remote ERROR    uncaught exception: ClassValidationError('While structuring Anchor', [AssertionError(), AssertionError()], <class 'fontra.core.classes.Anchor'>)
  + Exception Group Traceback (most recent call last):
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/remote.py", line 105, in _performCall
  |     returnValue = await methodHandler(*arguments, connection=self)
  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/fonthandler.py", line 290, in editFinal
  |     await self.updateLocalDataAndWriteToBackend(finalChange, connection)
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/fonthandler.py", line 320, in updateLocalDataAndWriteToBackend
  |     await self._updateLocalDataAndWriteToBackend(change, sourceConnection, False)
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/fonthandler.py", line 337, in _updateLocalDataAndWriteToBackend
  |     applyChange(rootObject, change)
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/changes.py", line 91, in applyChange
  |     _applyChange(subject, change)
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/changes.py", line 113, in _applyChange
  |     changeFunc(subject, *args, itemCast=itemCast)
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/changes.py", line 16, in setItem
  |     item = itemCast(item)
  |            ^^^^^^^^^^^^^^
  |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/classes.py", line 443, in structure
  |     return _cattrsConverter.structure(obj, cls)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/just/code/git/BlackFoundry/fontra/venv/lib/python3.12/site-packages/cattrs/converters.py", line 332, in structure
  |     return self._structure_func.dispatch(cl)(obj, cl)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "<cattrs generated structure fontra.core.classes.Anchor>", line 25, in structure_Anchor
  |     if errors: raise __c_cve('While structuring ' + 'Anchor', errors, __cl)
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | cattrs.errors.ClassValidationError: While structuring Anchor (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "<cattrs generated structure fontra.core.classes.Anchor>", line 10, in structure_Anchor
    |     res['x'] = __c_structure_x(o['x'], __c_type_x)
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/classes.py", line 325, in _structureNumber
    |     assert isinstance(d, (float, int))
    | AssertionError
    | Structuring class Anchor @ attribute x
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   File "<cattrs generated structure fontra.core.classes.Anchor>", line 15, in structure_Anchor
    |     res['y'] = __c_structure_y(o['y'], __c_type_y)
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/just/code/git/BlackFoundry/fontra/src/fontra/core/classes.py", line 325, in _structureNumber
    |     assert isinstance(d, (float, int))
    | AssertionError
    | Structuring class Anchor @ attribute y
    +------------------------------------
```

Reply:
The same happens when I edit a value in the "Add Anchor" dialog.

Reply:
I fixed the error message.